🍒 [windows][lldb] implement system logging on Windows (#150213)

charles-zablit opened the pull request:
- https://github.com/llvm/llvm-project/pull/150213
- https://github.com/llvm/llvm-project/pull/154424
- https://github.com/llvm/llvm-project/pull/156474

charles-zablit committed:
[windows][lldb] implement system logging on Windows (#150213)

This patch makes LLDB use the Event Viewer on Windows (equivalent of
system logging on Darwin) rather than piping to the standard output
(which was deactivated in ca0a5247004b6d692978d10bdbf86e338133e60c.

charles-zablit commented:
@swift-ci please test

charles-zablit commented:
@swift-ci please test

charles-zablit commented:
@swift-ci please test

charles-zablit commented:
@swift-ci please test windows

charles-zablit commented:
@swift-ci please test

charles-zablit commented:
@swift-ci please test Windows

charles-zablit commented:
@swift-ci please test windows

charles-zablit commented:
@swift-ci please test Windows

compnerd commented on lldb/source/Host/windows/Host.cpp:
```diff
@@ -302,3 +304,64 @@ Environment Host::GetEnvironment() {
+    event_type = EVENTLOG_INFORMATION_TYPE;
+  }
+
+  ReportEventW(h, event_type, 0, 0, nullptr, 1, 0, &msg_ptr, nullptr);
```
Is this really the best place to put the messages? The Event Log is for system events, this feels more developer oriented and feels like the kernel debug buffer might be better (`OutputDebugStringA`).

charles-zablit replied:
From my understanding, the messages outputted through `OutputDebugStringA` are only accessible from an attached debugger or something like DebugView (which still requires attaching). The goal of this PR is to mimic what `lldb` does on Darwin with system logging, which does not require a debugger.

> this feels more developer oriented

I think the health logs are useful for users as well to understand if their install is broken. I would say that they are also useful for lldb users.

compnerd replied:
Sure, but I don't think that the event viewer is the right location for the logging. This log tends to be very very very very verbose. If this only was triggered in the case that lldb is dying, that would be one thing. But this is effectively triggered on each run of lldb.

charles-zablit replied:
> This log tends to be very very very very verbose.

Are you talking about the overall system events or just the lldb health logs? I agree that the overall system logs (the contents of the event viewer) has a lot of entries. However the UI and the search bar allow to filter them easily. The Console on macOS also has a lot of entries and we can filter messages fine. I assume it's the same on Linux.

My reasoning for using the event viewer is:
1. Even though there are a lot of entries, the UI allows to quickly find `lldb` related entries.
2. This is the closest match to how we log on other platforms. We could switch to using `OutputDebugStringA` however we should also do so on other platforms. I don't think it's a good UX for users to need to use a debugger on Windows to get the health logs, but to only have to open the Console to get them on macOS.

I would be in favor of taking this change in `6.2` and to propose a migration to another destination for health logs on all platforms on the forums further down the line. What do you think?

compnerd replied:
I am referring to the lldb side. I don't think that Linux is the same. syslog is just a flat file though these days you would find the logs somewhere in systemd I imagine which isn't really easy to filter.

This is too late for 6.2 IMO. So we are better off just changing this now before merging.

compnerd commented on lldb/source/Host/windows/Host.cpp:
```diff
@@ -302,3 +304,64 @@ Environment Host::GetEnvironment() {
+  std::wstring unicode(unicode_length, L'\0');
+  MultiByteToWideChar(CP_ACP, 0, ansi.c_str(), -1, &unicode[0], unicode_length);
+  return unicode;
+}
```
Why re-implement `UTF8ToUTF16` and `UTF16ToUTF8` from LLVM?

charles-zablit replied:
Opened a PR upstream to address this. Thanks!
- https://github.com/llvm/llvm-project/pull/154424

charles-zablit committed:
[lldb][windows] remove duplicate implementation of UTF8ToUTF16 (#154424)

`std::wstring AnsiToUtf16(const std::string &ansi)` is a
reimplementation of `llvm::sys::windows::UTF8ToUTF16`. This patch
removes `AnsiToUtf16` and its usages entirely.

charles-zablit committed:
[lldb][windows] use OutputDebugStringA instead of c to log events (#156474)

In https://github.com/llvm/llvm-project/pull/150213 we made use of the
Event Viewer on Windows (equivalent of system logging on Darwin) rather
than piping to the standard output. This turned out to be too verbose in
practice, as the Event Viewer is developer oriented and not user
oriented.

This patch swaps the use of `ReportEventW` for `OutputDebugStringA`,
allowing to use tools such as `DebugView` to record logs when we are
interested in receiving them, rather than continuously writing to the
buffer. Please see an example below:
<img width="1253" height="215" alt="Screenshot 2025-09-02 at 16 07 03"
src="https://github.com/user-attachments/assets/4a326e46-d8a4-4c99-8c96-1bee62da8d55"
/>

charles-zablit commented:
@swift-ci please test

charles-zablit commented:
@swift-ci please test linux

charles-zablit commented:
@swift-ci please test macos

charles-zablit commented:
@swift-ci please test windows

charles-zablit commented:
CI is failing on Windows because of

```log
[2025-09-03T03:00:18.570Z] ********************
[2025-09-03T03:00:18.570Z] Unexpectedly Passed Tests (2):
[2025-09-03T03:00:18.570Z]   lldb-shell :: Settings/TestEchoCommands.test
[2025-09-03T03:00:18.570Z]   lldb-shell :: SymbolFile/DWARF/x86/dead-code-filtering.yaml
```

Restarting the CI as it's probably a flaky test on Windows.

@swift-ci please test windows

adrian-prantl commented:
@charles-zablit should we considered pulling this into 6.2.1?

charles-zablit commented:
> @charles-zablit should we considered pulling this into 6.2.1?

Reopened a PR here to avoid adding a lot of people as reviewers:
- https://github.com/swiftlang/llvm-project/pull/11315
